Seed finder - Gen 4

sorendog opened the pull request:
Tool for finding Initial seeds for gen 4 Method J and K.

Draft atm but I have the basic's started for method J.

To do list:

- copy and paste method j and replace division with modulo numbers for method k
- create tests
- make method j match pokefinder results

sorendog commented:
@zaksabeast When you get a chance would you mind taking a looking at the no lead method J I have started - the test is meant to fail atm as I am using it to get result to compare with poke finder - but when i compare nothing matches up and i can't seem to figure out why. I think I am doing this right, form pid first and then do a rng call for nature rand to see if match and if none, then i just have it return empty cause that means it isn't a valid seed. Which is backwards of how i coded the searcher static function.

Any help is greatly appreciated as always.

sorendog commented:
i'm gonna add a commit in a sec, realized i could code the missing loop i did in static. i don't think it is correct as i still can't get it to match pokefinder but like nature and gender match on the advance now it looks like.

sorendog commented:
Alright! This is ready to go. I think it might be a little rough around the edges, but it matches up with pokefinder results at least.

sorendog commented:
It's all good - I appreciate the heads ^^

sorendog commented:
I can't seem to figure out why method k synch isn't working - i thought maybe the wrong seed is being sent but when I check the results the test produces to pokefinder, it is either an advance off or doesn't exist even after making an adjustment for that thought. the rands are the same between j and k and i even checked pokefinder code in case maybe just k had it's own thing but i don't see anything for synch.

All that to say, I think i need help.

zaksabeast commented:
@sorendog I think I can point you in the right direction. There appear to be two issues.

## Expected results don't match PokeFinder

I pulled the latest changes and noticed `cargo test` fails for all the method k tests.

When I looked at `static_methodk`, I noticed the expected results don't line up with PokeFinder.

For example, the first expected test result is PID `2590bb5e` with IVs `30 / 30 / 20 / 24 / 31 / 27`, but PokeFinder shows `72C96FBE` with IVs `30 / 30 / 21 / 23 / 20 / 24`.

<img width="1318" height="898" alt="Screenshot 2025-07-30 at 7 14 30 PM" src="https://github.com/user-attachments/assets/ae05f289-8dca-496b-81b5-f2ed12d09f5c" />

However even then, the results produced by `search_static4_methodk_seeds` don't line up with PokeFinder either, so there's another issue somewhere.

## Tracking down the second issue

The steps we take to find the seed are:
- Recover the RNG state from IVs
- Recover PID from IVs
- Reverse the RNG state to find a seed with an acceptable date/time/delay/advance range

I walked through each step to track down where the issue is.

### recover_poke_rng_iv

I started by confirming `recover_poke_rng_iv` was finding the correct seed.

We know the IVs for the first PokeFinder result are `30 / 30 / 21 / 23 / 20 / 24`, so we can use that as the input. PokeFinder shows `A2170321` is the initial seed, so that's our eventual expected result.

I wrote a quick debug test to recover the RNG states:
```rs
#[test]
fn ztest() {
    let opts = SearchStatic4MethodjOpts {
        // opts from static_methodk test
    };

    let target_ivs = ivs!(30 / 30 / 21 / 23 / 20 / 24);

    let result = recover_poke_rng_iv(&target_ivs, false);
    assert_eq!(result, [])
}
```

You can run a single test by adding it to the end of the command: `cargo test -- ztest`. This makes it fast to run one-off debug tests.

The result it gave was:
```
[ 0x57de3db7, 0xd7de3db7, 0x57dea58a, 0xd7dea58a, ]
```

Since these RNG states are reversed to the get initial seed of `A2170321`, I plugged each of these into PokeFinder's researcher tool with LCRNG[R] and searched for `A2170321`.

`57de3db7` had it at 117:

<img width="1090" height="810" alt="Screenshot 2025-07-30 at 7 33 36 PM" src="https://github.com/user-attachments/assets/9534861e-ccec-4bc2-9151-af91fc67bd9e" />

So `recover_poke_rng_iv` is correct 👍 

### get_state_from_kseed

Next we recover the PID using `get_state_from_kseed`. From the previous test we know the RNG state should be `57de3db7`, so I used that as an input.

```rs
#[test]
fn ztest() {
    let opts = SearchStatic4MethodjOpts {
        // opts from static_methodk test
    };

    let target_ivs = ivs!(30 / 30 / 21 / 23 / 20 / 24);
    let result = get_state_from_kseed(&opts, target_ivs, 0x57de3db7);
    assert_eq!(result, []);    
}
```

The results were:
```
[
    { seed: 0xced3098f, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
    { seed: 0x439b25a7, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
    { seed: 0x772c4193, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
    { seed: 0x5b81c665, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 }
]
```

Each of these has the PID that matches PokeFinder, so we know the PID is being recovered successfully.

Since the settings for the `static_methodk` test ask for a seed that would produce the Pokemon between 20-21 advances, one of these new RNG states, `CED3098F, 439B25A7, 772C4193, 5B81C665`, must reverse back to `A2170321` within just a few advances.

And sure enough, `5B81C665` does that (researcher appears to have an off-by-1 bug)

<img width="1090" height="810" alt="Screenshot 2025-07-30 at 7 39 20 PM" src="https://github.com/user-attachments/assets/8c77d091-8017-49e2-8b03-95ac04e57af6" />

So we've confirmed `get_state_from_kseed` works as expected 👍 

### search_static4_methodk_seeds

At this point we know we're correctly recovering the PID and RNG states, and we've so far produced the expected result.

The last thing to check is reversing `5B81C665` back to `A2170321` and getting a seed time.

Running `search_static4_methodk_seeds` produces a lot of results, so to reduce the noise I changed my test's tid and sid to make the expected PID `72C96FBE` shiny, then I enabled the shiny filter:
```rs
#[test]
fn ztest() {
    let opts = SearchStatic4MethodjOpts {
        // opts from static_methodk test
    };

    opts.tid = 0x72C9;
    opts.sid = 0x6FBE;
    opts.filter.shiny = true;

    let result = search_static4_methodk_seeds(&opts);
    assert_eq!(result, [])
}
```

This meant I should only find the single result we've been targeting.

I added a log in the loop reversing the RNG state to let me know the `5B81C665` state was making it that far:
```rs
for advance in min_advance..=max_advance {
    if state.seed == 0x5b81c665 {
        // New log
        println!("0x5b81c665 reversed to {:08x}", seed);
    }
```

You can see logs in rust tests by adding `--nocapture` to the command: `cargo test -- --nocapture ztest`.

Sure enough, I see this log:
```
0x5b81c665 reversed to a2170321
```

Which confirms we _are_ reversing back to `a2170321`! That means it must be getting filtered out somewhere later.

I added another log to see the seed time that was produced:
```rs
let seed_time = hgss_calculate_seedtime(seed_time_opts);

if seed == 0xa2170321 {
    println!("0xa2170321 seed_time {:?}", seed_time);
}
```

But this time the log was:
```
0xa2170321 seed_time []
```

That means `hgss_calculate_seedtime` is not be calculating the correct seed time, or the inputs are incorrect.

## Next steps

I would write a test for `hgss_calculate_seedtime` to find a seed time for `a2170321` was the same inputs that are given in the `search_static4_methodk_seeds`. That should help uncover why the a seed time isn't found.

sorendog commented:
> @sorendog I think I can point you in the right direction. There appear to be two issues.
> 
> ## Expected results don't match PokeFinder
> I pulled the latest changes and noticed `cargo test` fails for all the method k tests.
> 
> When I looked at `static_methodk`, I noticed the expected results don't line up with PokeFinder.
> 
> For example, the first expected test result is PID `2590bb5e` with IVs `30 / 30 / 20 / 24 / 31 / 27`, but PokeFinder shows `72C96FBE` with IVs `30 / 30 / 21 / 23 / 20 / 24`.
> 
> <img alt="Screenshot 2025-07-30 at 7 14 30 PM" width="1318" height="898" src="https://private-user-images.githubusercontent.com/19464008/472761565-ae05f289-8dca-496b-81b5-f2ed12d09f5c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQxNjMwMjMsIm5iZiI6MTc1NDE2MjcyMywicGF0aCI6Ii8xOTQ2NDAwOC80NzI3NjE1NjUtYWUwNWYyODktOGRjYS00OTZiLTgxYjUtZjJlZDEyZDA5ZjVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAyVDE5MjUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljYmM3ZmJhZWRkNDUyODNjM2U3YzMxYjNjZDY1YWUzZTMwNjNjYmNmOWNlYjFhZTgyM2Q4ZWNiYTlkNTgzMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cts6yaXg0qt6hoLNvgBMBUcrOKrahHBySWkg6acm8eI">
> However even then, the results produced by `search_static4_methodk_seeds` don't line up with PokeFinder either, so there's another issue somewhere.
> 
> ## Tracking down the second issue
> The steps we take to find the seed are:
> 
> * Recover the RNG state from IVs
> * Recover PID from IVs
> * Reverse the RNG state to find a seed with an acceptable date/time/delay/advance range
> 
> I walked through each step to track down where the issue is.
> 
> ### recover_poke_rng_iv
> I started by confirming `recover_poke_rng_iv` was finding the correct seed.
> 
> We know the IVs for the first PokeFinder result are `30 / 30 / 21 / 23 / 20 / 24`, so we can use that as the input. PokeFinder shows `A2170321` is the initial seed, so that's our eventual expected result.
> 
> I wrote a quick debug test to recover the RNG states:
> 
> ```rust
> #[test]
> fn ztest() {
>     let opts = SearchStatic4MethodjOpts {
>         // opts from static_methodk test
>     };
> 
>     let target_ivs = ivs!(30 / 30 / 21 / 23 / 20 / 24);
> 
>     let result = recover_poke_rng_iv(&target_ivs, false);
>     assert_eq!(result, [])
> }
> ```
> 
> You can run a single test by adding it to the end of the command: `cargo test -- ztest`. This makes it fast to run one-off debug tests.
> 
> The result it gave was:
> 
> ```
> [ 0x57de3db7, 0xd7de3db7, 0x57dea58a, 0xd7dea58a, ]
> ```
> 
> Since these RNG states are reversed to the get initial seed of `A2170321`, I plugged each of these into PokeFinder's researcher tool with LCRNG[R] and searched for `A2170321`.
> 
> `57de3db7` had it at 117:
> 
> <img alt="Screenshot 2025-07-30 at 7 33 36 PM" width="1090" height="810" src="https://private-user-images.githubusercontent.com/19464008/472764469-9534861e-ccec-4bc2-9151-af91fc67bd9e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQxNjMwMjMsIm5iZiI6MTc1NDE2MjcyMywicGF0aCI6Ii8xOTQ2NDAwOC80NzI3NjQ0NjktOTUzNDg2MWUtY2NlYy00YmMyLTkxNTEtYWY5MWZjNjdiZDllLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAyVDE5MjUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1MDk5NGJiNGRmNDg5YTMzZDVjYWIyZTEwOWM1ZDNlNTg1MGRiMTUwY2Y5OWEzNWU0NmM5OWJlZWM4MDY1YjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.h7mOjqtfr3RFCIETWQ0rOth0uX69FAj3Tl9VTWBwG6g">
> So `recover_poke_rng_iv` is correct 👍
> 
> ### get_state_from_kseed
> Next we recover the PID using `get_state_from_kseed`. From the previous test we know the RNG state should be `57de3db7`, so I used that as an input.
> 
> ```rust
> #[test]
> fn ztest() {
>     let opts = SearchStatic4MethodjOpts {
>         // opts from static_methodk test
>     };
> 
>     let target_ivs = ivs!(30 / 30 / 21 / 23 / 20 / 24);
>     let result = get_state_from_kseed(&opts, target_ivs, 0x57de3db7);
>     assert_eq!(result, []);    
> }
> ```
> 
> The results were:
> 
> ```
> [
>     { seed: 0xced3098f, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
>     { seed: 0x439b25a7, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
>     { seed: 0x772c4193, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 },
>     { seed: 0x5b81c665, pid: 0x72c96fbe, ivs: Ivs { hp: 30, atk: 30, def: 21, spa: 23, spd: 20, spe: 24 }, ability: First, gender: Male, nature: Mild, shiny: false, characteristic: LovesToEat, advance: 0 }
> ]
> ```
> 
> Each of these has the PID that matches PokeFinder, so we know the PID is being recovered successfully.
> 
> Since the settings for the `static_methodk` test ask for a seed that would produce the Pokemon between 20-21 advances, one of these new RNG states, `CED3098F, 439B25A7, 772C4193, 5B81C665`, must reverse back to `A2170321` within just a few advances.
> 
> And sure enough, `5B81C665` does that (researcher appears to have an off-by-1 bug)
> 
> <img alt="Screenshot 2025-07-30 at 7 39 20 PM" width="1090" height="810" src="https://private-user-images.githubusercontent.com/19464008/472765371-8c77d091-8017-49e2-8b03-95ac04e57af6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQxNjMwMjMsIm5iZiI6MTc1NDE2MjcyMywicGF0aCI6Ii8xOTQ2NDAwOC80NzI3NjUzNzEtOGM3N2QwOTEtODAxNy00OWUyLThiMDMtOTVhYzA0ZTU3YWY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAyVDE5MjUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwNzM1NDMxNDZlM2YyMmExMzM1ZmUzMDYwNDljNTk3ZWQ0Y2RkYWU1MjZiNDkzOTQ1ZWNlNTUxOGVmMzE2YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M9kJyELtSqCeJrBYsqMYgrO-FvdYaSdQjSR8e5mVX1Y">
> So we've confirmed `get_state_from_kseed` works as expected 👍
> 
> ### search_static4_methodk_seeds
> At this point we know we're correctly recovering the PID and RNG states, and we've so far produced the expected result.
> 
> The last thing to check is reversing `5B81C665` back to `A2170321` and getting a seed time.
> 
> Running `search_static4_methodk_seeds` produces a lot of results, so to reduce the noise I changed my test's tid and sid to make the expected PID `72C96FBE` shiny, then I enabled the shiny filter:
> 
> ```rust
> #[test]
> fn ztest() {
>     let opts = SearchStatic4MethodjOpts {
>         // opts from static_methodk test
>     };
> 
>     opts.tid = 0x72C9;
>     opts.sid = 0x6FBE;
>     opts.filter.shiny = true;
> 
>     let result = search_static4_methodk_seeds(&opts);
>     assert_eq!(result, [])
> }
> ```
> 
> This meant I should only find the single result we've been targeting.
> 
> I added a log in the loop reversing the RNG state to let me know the `5B81C665` state was making it that far:
> 
> ```rust
> for advance in min_advance..=max_advance {
>     if state.seed == 0x5b81c665 {
>         // New log
>         println!("0x5b81c665 reversed to {:08x}", seed);
>     }
> ```
> 
> You can see logs in rust tests by adding `--nocapture` to the command: `cargo test -- --nocapture ztest`.
> 
> Sure enough, I see this log:
> 
> ```
> 0x5b81c665 reversed to a2170321
> ```
> 
> Which confirms we _are_ reversing back to `a2170321`! That means it must be getting filtered out somewhere later.
> 
> I added another log to see the seed time that was produced:
> 
> ```rust
> let seed_time = hgss_calculate_seedtime(seed_time_opts);
> 
> if seed == 0xa2170321 {
>     println!("0xa2170321 seed_time {:?}", seed_time);
> }
> ```
> 
> But this time the log was:
> 
> ```
> 0xa2170321 seed_time []
> ```
> 
> That means `hgss_calculate_seedtime` is not be calculating the correct seed time, or the inputs are incorrect.
> 
> ## Next steps
> I would write a test for `hgss_calculate_seedtime` to find a seed time for `a2170321` was the same inputs that are given in the `search_static4_methodk_seeds`. That should help uncover why the a seed time isn't found.

So - I was looking this over. the result produced for static is something i am able to find within pokefinder. Because of how pokefinder looks for delay, if you do the same delay range as this tool, it gets filtered out but by increasing the range to 800 and 830 within pokefinder it is the first result. I've been using the generator tab to confirm test results because of this. The test fails currently, because i didn't put the valid results for the roamer locations or the elm calls yet for all of static k. Which I did forget I had done, I'm sorry for not mentioning it! I will go through the results however and double check cute charm as well.
<img width="971" height="630" alt="image" src="https://github.com/user-attachments/assets/f918698b-15aa-4d51-904a-b58cacc80702" />

sorendog commented:
Only the first result matches, the second on static is one off now. I will review the time finder for hgss and see if that is whats causing the issue.

sorendog commented on rng_tools/src/generators/gen4/stationary/searcher.rs:
```diff
@@ -571,6 +613,64 @@ pub fn search_static4_methodj_seeds(
 }
 
+fn search_static4_methodk(opts: &SearchStatic4MethodjOpts) -> Vec<Base4MethodjState> {
+    let Ivs {
+        hp: min_hp,
+        atk: min_atk,
+        def: min_def,
+        spa: min_spa,
+        spd: min_spd,
+        spe: min_spe,
+    } = opts.filter.min_ivs;
+
+    let Ivs {
+        hp: max_hp,
+        atk: max_atk,
+        def: max_def,
+        spa: max_spa,
+        spd: max_spd,
+        spe: max_spe,
+    } = opts.filter.max_ivs;
+
+    iproduct!(
+        min_hp..=max_hp,
+        min_atk..=max_atk,
+        min_def..=max_def,
+        min_spa..=max_spa,
+        min_spd..=max_spd,
+        min_spe..=max_spe
+    )
+    .flat_map(|(hp, atk, def, spa, spd, spe)| {
```
That makes sense, this logic will be used for pretty much all searchers like this won't it? I'll note it down so if i don't come up with something else, maybe this and some other small things can be done all together as a PR in the future

zaksabeast reviewed (approved):
Thank you!
